Add high-level documentation about Java traces [DOC-128]

jeannielynnmoulton opened the pull request:
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- ~[] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).~ not applicable
- ~[] My commit message includes data points confirming performance improvements (if claimed).~ not applicable
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

Commit:
Add high-level documentation about Java traces.

This overview is intended to give an overview of what a user might see if they inspect a trace of the execution of a Java program. It details function calls, function returns and gives examples assignments, but notin detail what the expressions on each side of the assignments look like.

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 0e696e842f8921145b1d26d766aeaa569163cd8a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/91254685
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: e5d9bd97d9c6e6e156b5aaadfce9f820461e88f6).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/91255414

tautschnig commented on src/goto-programs/README.md:
```diff
@@ -509,3 +509,80 @@ C++ modules:
   - [linking folder](\ref linking).
   - [typecheck](\ref section-goto-typecheck).
+
+\section java-trace-structure Java Trace Structure
```
I'm not sure I agree that talking about Java should happen in this folder. Looking at the contents, however, it doesn't particularly Java-specific anyway. Can the title and the introduction below be adjusted a bit to make clear that it's only about the example being in Java, which seems perfectly fine to me?

peterschrammel replied:
Shouldn't this be called 'Goto Trace Structure' ?
A goto trace corresponds to a goto program in the first place; the link to C or Java is an interpretation of it... So, I'd start with the generic description and then add separate sections to detail language-specific interpretations of it.

peterschrammel commented on src/goto-programs/README.md:
```diff
@@ -509,3 +509,80 @@ C++ modules:
+- function call
+- function return
+- assignment
+
```
assert?

jeannielynnmoulton replied:
This is mentioned below.

peterschrammel commented on src/goto-programs/README.md:
```diff
@@ -509,3 +509,80 @@ C++ modules:
+
+The expressions on each side vary depending on the Java type, whether its a
+member assignment, whether its an assignment by reference, etc. Some examples:
+- A global variable assignment will have a LHS \ref symbolt containing the
```
The Java specific section should start here.

thk123 replied:
:pick: `\ref symbol_exprt`

peterschrammel commented on src/goto-programs/README.md:
```diff
@@ -509,3 +509,80 @@ C++ modules:
+(\ref goto_trace_stept::full_lhs) and the value expression representing the RHS 
+(\ref goto_trace_stept::full_lhs_value).
+
+The expressions on each side vary depending on the Java type, whether its a
```
The same is true for plain GOTO (which has structs, arrays, etc).

peterschrammel commented on src/goto-programs/README.md:
```diff
@@ -509,3 +509,80 @@ C++ modules:
+
+In a trace of a Java program, there are other step types that are internal
+including:
+- goto_trace_stept::typet::ASSERT (failure), used in bytecode instrumentation
```
...and properties

peterschrammel reviewed (commented):
I'd like to see this explained wrt 'plain' GOTO first. The Java-specific things should be in a separate section or preferably even in JBMC, e.g. jbmc/src/java_bytecode/README.md

smowton commented on src/goto-programs/README.md:
```diff
@@ -509,3 +509,80 @@ C++ modules:
+call. The type of a step can be checked using the boolean functions such as 
+\ref goto_trace_stept::is_function_call(). 
+
+The remainder of the members of goto_tracet depend on the step type.
```
The meaning of the remainder of the members of `goto_tracet_stept`?

smowton commented on src/goto-programs/README.md:
```diff
@@ -509,3 +509,80 @@ C++ modules:
+The expressions on each side vary depending on the Java type, whether its a
+member assignment, whether its an assignment by reference, etc. Some examples:
+- A global variable assignment will have a LHS \ref symbolt containing the
+Java type information and an identifier that does not include a function
```
Perhaps just an identifier corresponding to a symbol with `is_static_lifetime` set?

thk123 commented on src/goto-programs/README.md:
```diff
@@ -509,3 +509,80 @@ C++ modules:
+and signifies that the program is calling a function.
+The step contains a valid identifier of the function being called 
+(\ref goto_trace_stept::called_function) and a vector of the function arguments 
+(\ref goto_trace_stept::function_arguments).
```
:pick: Mention that `\ref goto_trace_stept::function` is the function of the call site

thk123 commented on src/goto-programs/README.md:
```diff
@@ -509,3 +509,80 @@ C++ modules:
+A function return step has the goto_trace_stept::typet 
+\ref goto_trace_stept::typet::FUNCTION_RETURN and signifies that the program is 
+exiting a function. There is no special information contained in a function 
+return step.
```
:pick: Mention that `\ref goto_trace_stept::function` is the function being returned from.

thk123 reviewed (approved):
Nothing blocking (though splitting the Java specific stuff out would be good). Perhaps in a follow up doc day we should explain about some of the multi-step structure (like how a function call is followed by assignments to each of the parameters).

jeannielynnmoulton commented:
@peterschrammel I've addressed your comments, can you please re-review (there's a separate commit for convenience)? @thk123 I am working on adding some docs for more complex objects but I would like to do those in another PR and perhaps in a different location to keep things moving.

thk123 commented on src/goto-programs/README.md:
```diff
@@ -534,3 +534,101 @@ C++ modules:
+
+A function return step has the goto_trace_stept::typet 
+\ref goto_trace_stept::typet::FUNCTION_RETURN and signifies that the program is 
+exiting a function. \ref goto_trace_stept::function is the function being 
```
This becomes `goto_trace_stept::function_id` as doxygen error has pointed out.

peterschrammel reviewed (approved):
CI is still failing.
Also, please squash commits.

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: bcee5b328bad78a7df0512c43886ae71c601fc21).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99838363
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 4dc4b65441f3026f2a3530317da3afd5826bb711).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99838708
